raft/raft.go

```diff
@@ -0,0 +1,292 @@
+			log.WithFields(log.Fields{}).Infof("reset election timer")
+			// reset election timer
+			r.resetElectionTimer()
+			// do another election
+			electionChan <- struct{}{}
+		case <-r.appendEntriesSignal:
+			log.WithFields(log.Fields{}).Infof("heartbeat received")
+			r.state.Lock()
+			r.state.ChangeRole(model.RaftRoleFollower)
+			r.state.VotedFor = ""
+			r.state.Unlock()
+			return
+		case <-electionChan:
+			log.WithFields(log.Fields{}).Infof("do election")
+			shouldReturn := r.doElection()
+			if shouldReturn {
+				return
+			}
+		}
+	}
+}
+
+func (r *defaultRaft) doElection() bool {
+	log.WithFields(log.Fields{}).Infof("do election start")
+
+	r.state.Lock()
+	defer r.state.Unlock()
+
+	// 1. increment current term
+	r.state.CurrentTerm++
+	// 2. vote for self
+	r.state.VotedFor = r.state.ID
+	// 3. reset election timer
+	r.resetElectionTimer()
+	// calculate majority:
+	majority := math.Ceil(float64(len(r.config.ClusterMemberAddreses)) / float64(2))
+	// 1 because we voted for ourselves
+	var votes int32 = 1
+
+	// Do voting
+	log.WithFields(log.Fields{}).Infof("do voting")
+	requestVoteReq := model.NewRequestVoteRequestDTO(
+		r.state.CurrentTerm,
+		r.state.ID,
+		int32(len(r.state.Logs)),
+		0)
+
+	voteRespChan := make(chan *model.RequestVoteResponseDTO, len(r.config.ClusterMemberAddreses))
+
+	var wg sync.WaitGroup
+	for _, memberAddress := range r.config.ClusterMemberAddreses {
+		if memberAddress != r.address {
+			wg.Add(1)
+			go func(memberAddress string) {
+				defer wg.Done()
+				resp, err := r.raftClient.RequestVote(memberAddress, requestVoteReq)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"err": err,
+					}).Error("error requesting vote")
+					return
+				}
+
+				log.WithFields(log.Fields{
+					"id":      r.state.ID,
+					"address": memberAddress,
+					"resp":    resp,
+				}).Infof("do voting done")
+				voteRespChan <- resp
+			}(memberAddress)
+		}
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		voteSeen := 0
+		for voteResp := range voteRespChan {
+			voteSeen++
+			if voteResp.VoteGranted {
+				atomic.AddInt32(&votes, 1)
+			}
+			if voteSeen == len(r.config.ClusterMemberAddreses)-1 {
+				return
+			}
+		}
+	}()
+
+	wg.Wait()
+
+	log.WithFields(log.Fields{
+		"id":             r.state.ID,
+		"voting_result":  votes,
+		"majority":       majority,
+		"will_be_leader": votes >= int32(majority),
+		"term":           r.state.CurrentTerm,
+	}).Infof("voting result")
+
+	if votes >= int32(majority) {
+		r.state.ChangeRole(model.RaftRoleLeader)
+		return true
+	}
+
+	return false
+}
+
+func (r *defaultRaft) doLeaderAction() {
+	log.WithFields(log.Fields{}).Infof("do leader action start")
+	r.heartbeatTicker = time.NewTicker(r.config.HeartbeatInterval)
+
+	for range r.heartbeatTicker.C {
+		for _, memberAddress := range r.config.ClusterMemberAddreses {
+			if memberAddress != r.address {
+				go r.raftClient.AppendEntries(memberAddress, model.NewAppendEntriesRequestDTO())
```

Comment:
cmiiw. sepahaman gw heartbeat ini cuma perlu utk kondisi idle ketika gak ada entries dari client, karena ketika ada write dari client, itu juga bisa dipakai sebagai pengganti healthcheck

Reply:
hmm kalo dari visualisasi di https://raft.github.io kok kayaknya itu tetep kirim heartbeat ya?

Reply:
sebenernya agak susah ngeobserve ini dari visualisasi. tapi kalau requestnya clientnya cukup frequent, sebenernya heartbeat jadi gak perlu. karena payload append entries rpc utk ngecommit entry yg baru masuk tetep bisa dipake sama receivernya.

Dari paper raft:
> Upon election: send initial empty AppendEntries RPCs
(heartbeat) to each server; repeat during idle periods to
prevent election timeouts (§5.2)

Reply:
oh right

jadi paling bikin idle timeout aja ya instead of heartbeat

thanks2